Specifications/QIR/Data-Types.md

```diff
@@ -272,7 +297,7 @@ arrays:
 |----------------------------------|--------------------------------------|-------------|
 | __quantum__rt__array_create_1d   | `%Array* void(i32, i64)`             | Creates a new 1-dimensional array. The `i32` is the size of each element in bytes. The `i64` is the length of the array. The bytes of the new array should be set to zero. If the length is zero, the result should be an empty 1-dimensional array. |
 | __quantum__rt__array_copy        | `%Array*(%Array*)`                   | Returns a new array which is a copy of the passed-in `%Array*`. |
-| __quantum__rt__array_concatenate | `%Array*(%Array*, %Array*)`          | Returns a new array which is the concatenation of the two passed-in arrays. |
+| __quantum__rt__array_concatenate | `%Array*(%Array*, %Array*)`          | Returns a new array which is the concatenation of the two passed-in one-dimensional arrays. If either array is not one-dimensional, then a runtime failure should occur. |
```

Comment:
The arrays also must contain elements of the same size.
Also, is it allowed to concatenate arrays of qubits? If so, how the arrays/qubits should be released?

Reply:
Yes, arrays of qubits can be concatenated. I'd like to understand the question around how they should be released better. What is the relation between array concatenation and releasing?

Reply:
When creating a new array of qubits, one must then release it with `__quantum__rt__qubit_release_array`. The question is, should this call release the qubits the array contains or is it the client's responsibility to release the qubits manually and only after that release the array itself? So far I've been assuming the former (otherwise, why provide a separate array alloc/release for arrays of qubits?). In either case, the description of `__quantum__rt__qubit_release_array` should probably be clarified.

Reply:
Thinking some more about it, `__quantum__rt__qubit_release_array` seems to be just syntactic sugar over 1) release all qubits it contains, 2) release the array -- given how easy it is to cause double release of qubits by using it (if we allow arrays of qubits to be copied, concatenated, etc.) -- should it even be provided?

Arguably, `__quantum__rt__qubit_allocate_array` is also just a shortcut, but it's safe so I don't have strong feelings against providing it.

Reply:
I think the rule is that any array that is created with `__quantum__rt__qubit_allocate_array` must be released with `__quantum__rt__qubit_release_array` and no array created in any other way may be released with `__quantum__rt__qubit_release_array`. An array of qubits you create by concatenating two existing arrays of qubits should be unreferenced like any other array.

I'm reluctant to drop `__quantum__rt__qubit_release_array` because replacing it with an LLVM loop will make the QIR even more verbose than it already is. Also, it will encourage people to think that it's OK to allocate an array of qubits and then release them one by one at different times; the intent of allocating and releasing an array is that you're treating the qubits as a group (a quantum register), not as just a pool of random qubits.

It is certainly true that the current text doesn't make this clear. If you're both OK with what I just wrote, I'll add it to the text.

Reply:
To me, it feels counter-intuitive that concatenating two arrays of qubits creates a new array that is managed in a different way than the original arrays. The crossing between arrays of qubits and arrays of other types is already precarious. Would it make sense to have a separate type for arrays of qubits that owns its content (then concatenate, copy, etc. could take arrays of qubits but return non-managing array of `%Array*`)?

Re treating an array of qubits as a register: I'm fine with this, but what it would mean then to copy a register or to concatenate two registers?

Reply:
I think making a type distinction is likely to cause trouble. A thing that might be ok is if an array has a flag indicating just that. Even so, I am doubtful it is worth the hassle. To me, this has little to do with the array - allocation and deallocation is not associated with a specific data type. Instead, if/once we have initializers that e.g. allocate user defined types, then we might want to generate a method to allocate that type (calling into __quantum__rt__qubit_allocate_ / __quantum__rt__qubit_allocate_array as needed) and the corresponding deallocate method with it. 
I agree with keeping __quantum__rt__qubit_release_array.

Reply:
I added the requirement that the arrays have the same size. The rest is covered in Bettina's coming PR.